src/Simulation/Simulators/OpenSystemsSimulator/Decompositions/IntrinsicInterface.qs

```diff
@@ -0,0 +1,212 @@
+                    H(target);
+                } apply {
+                    Controlled Z(controls, target);
+                }
+            } else {
+                ApplyWithLessControlsA(Controlled X, (controls, target));
+            }
+        }
+    }
+
+    operation Y(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.Y(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                Y(target);
+            } elif Length(controls) == 1 {
+                within {
+                    Native.H(target);
+                    Native.S(target);
+                } apply {
+                    Native.CNOT(controls[0], target);
+                }
+            } elif Length(controls) == 2 {
+                within {
+                    within {
+                        Native.H(target);
+                    } apply {
+                        Native.S(target);
+                    }
+                } apply {
+                    Controlled Z(controls, target);
+                }
+            } else {
+                ApplyWithLessControlsA(Controlled Y, (controls, target));
+            }
+        }
+    }
+
+    operation Z(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.Z(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                Native.Z(target);
+            } elif Length(controls) == 1 {
+                within {
+                    H(target);
+                } apply {
+                    Native.CNOT(controls[0], target);
+                }
+            } elif Length(controls) == 2 {
+                // [Page 15 of arXiv:1206.0758v3](https://arxiv.org/pdf/1206.0758v3.pdf#page=15)
+                Adjoint Native.T(controls[0]);
+                Adjoint Native.T(controls[1]);
+                Controlled Native.X([target], controls[0]);
+                Native.T(controls[0]);
+                Native.CNOT(controls[1], target);
+                Native.CNOT(controls[1], controls[0]);
+                Native.T(target);
+                Adjoint Native.T(controls[0]);
+                Native.CNOT(controls[1], target);
+                Native.CNOT(target, controls[0]);
+                Adjoint Native.T(target);
+                Native.T(controls[0]);
+                Native.CNOT(controls[1], controls[0]);
+            } else {
+                ApplyWithLessControlsA(Controlled Z, (controls, target));
+            } 
+        }
+    }
+
+    operation S(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.S(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                Native.S(target);
+            } elif (Length(controls) == 1) {
+                Native.T(controls[0]);
+                Native.T(target);
+                within {
+                    Controlled X([controls[0]], target);
```

Comment:
This spot could also become `Native.CNOT`.

Reply:
Also fixed in 3591151c7c3ba48ec98ffbdd6ecfde9a254e4f33.